src/mono/wasm/runtime/driver.c

```diff
@@ -867,57 +869,67 @@ mono_wasm_marshal_type_from_mono_type (int mono_type, MonoClass *klass, MonoType
 	case MONO_TYPE_STRING:
 		return MARSHAL_TYPE_STRING;
 	case MONO_TYPE_SZARRAY:  { // simple zero based one-dim-array
-		MonoClass *eklass = mono_class_get_element_class (klass);
-		MonoType *etype = mono_class_get_type (eklass);
-
-		switch (mono_type_get_type (etype)) {
-			case MONO_TYPE_U1:
-				return MARSHAL_ARRAY_UBYTE;
-			case MONO_TYPE_I1:
-				return MARSHAL_ARRAY_BYTE;
-			case MONO_TYPE_U2:
-				return MARSHAL_ARRAY_USHORT;
-			case MONO_TYPE_I2:
-				return MARSHAL_ARRAY_SHORT;
-			case MONO_TYPE_U4:
-				return MARSHAL_ARRAY_UINT;
-			case MONO_TYPE_I4:
-				return MARSHAL_ARRAY_INT;
-			case MONO_TYPE_R4:
-				return MARSHAL_ARRAY_FLOAT;
-			case MONO_TYPE_R8:
-				return MARSHAL_ARRAY_DOUBLE;
-			default:
-				return MARSHAL_TYPE_OBJECT;
+		if (klass) {
+			MonoClass *eklass = mono_class_get_element_class (klass);
+			MonoType *etype = mono_class_get_type (eklass);
+
+			switch (mono_type_get_type (etype)) {
+				case MONO_TYPE_U1:
+					return MARSHAL_ARRAY_UBYTE;
+				case MONO_TYPE_I1:
+					return MARSHAL_ARRAY_BYTE;
+				case MONO_TYPE_U2:
+					return MARSHAL_ARRAY_USHORT;
+				case MONO_TYPE_I2:
+					return MARSHAL_ARRAY_SHORT;
+				case MONO_TYPE_U4:
+					return MARSHAL_ARRAY_UINT;
+				case MONO_TYPE_I4:
+					return MARSHAL_ARRAY_INT;
+				case MONO_TYPE_R4:
+					return MARSHAL_ARRAY_FLOAT;
+				case MONO_TYPE_R8:
+					return MARSHAL_ARRAY_DOUBLE;
+				default:
+					return MARSHAL_TYPE_OBJECT;
+			}
+		} else {
+			return MARSHAL_TYPE_OBJECT;
 		}
 	}
 	default:
 		mono_wasm_ensure_classes_resolved ();
 
-		if (klass == datetime_class)
-			return MARSHAL_TYPE_DATE;
-		if (klass == datetimeoffset_class)
-			return MARSHAL_TYPE_DATEOFFSET;
-		if (uri_class && mono_class_is_assignable_from(uri_class, klass))
-			return MARSHAL_TYPE_URI;
-		if (klass == voidtaskresult_class)
-			return MARSHAL_TYPE_VOID;
-		if (mono_class_is_enum (klass))
-			return MARSHAL_TYPE_ENUM;
-		if (!mono_type_is_reference (type)) //vt
-			return MARSHAL_TYPE_VT;
-		if (mono_class_is_delegate (klass))
-			return MARSHAL_TYPE_DELEGATE;
-		if (class_is_task(klass))
-			return MARSHAL_TYPE_TASK;
-		if (safehandle_class && (klass == safehandle_class || mono_class_is_subclass_of(klass, safehandle_class, 0))) {
-			return MARSHAL_TYPE_SAFEHANDLE;
+		if (klass) {
+			if (uri_class && mono_class_is_assignable_from(uri_class, klass))
+				return MARSHAL_TYPE_URI;
+			if (klass == voidtaskresult_class)
+				return MARSHAL_TYPE_VOID;
+			if (mono_class_is_enum (klass))
+				return MARSHAL_TYPE_ENUM;
+			if (type && !mono_type_is_reference (type)) //vt
+				return MARSHAL_TYPE_VT;
+			if (mono_class_is_delegate (klass))
+				return MARSHAL_TYPE_DELEGATE;
+			if (class_is_task(klass))
```

Comment:
Do we want to keep Task to be special and what about ValueTask?

Reply:
Task still needs to be special, but good point about ValueTask. I will find out whether ValueTask works and whether we need to fix it

Reply:
ValueTask<T> turned out to hit an obscure bug in driver.c that was flushed out by writing a test for it. I fixed the bug, but making ValueTask marshal properly will be a sizable amount of work so I think we want to do that later - perhaps with the SafeHandle changes. I definitely do not think ValueTask ever worked.